lib/Borel/Types/UOM.hs

```
@@ -0,0 +1,225 @@
+pUOM = prism' pretty parse
+  where dash = "-"
+        pretty (u `Times` v) = (u ^. re pUOM) <> dash <> (v ^. re pUOM)
+        pretty (UOM p b)     = (p ^. re pPrefixUOM) <> (b ^. re pBaseUOM)
+
+        -- treat UOM @Times@ as left-associative.
+        parse = hush . AT.parseOnly (parser <* AT.endOfInput)
+        parser = do
+          uoms <- puom `AT.sepBy` AT.string dash
+          case uoms of []     -> mzero
+                       -- construct left-associative UOMs
+                       (u:us) -> return $ foldl Times u us
+        puom = do
+          pre  <- AT.option (Just Base) ppre
+          base <- pbase
+          case (pre, base) of (Just p, Just b)  -> return $ UOM p b
+                              _                 -> mzero
+        ppre   =  (preview pPrefixUOM <$> AT.string "G")
+              <|> (preview pPrefixUOM <$> AT.string "n")
+              <|> (preview pPrefixUOM <$> AT.string "Mi")
+              <|> (preview pPrefixUOM <$> AT.string "M")
+        pbase =   (preview pBaseUOM <$> AT.string "s")
+              <|> (preview pBaseUOM <$> AT.string "h")
+              <|> (preview pBaseUOM <$> AT.string "B")
+              <|> (preview pBaseUOM <$> AT.string "instance")
+              <|> (preview pBaseUOM <$> AT.string "ip-address")
+              <|> (preview pBaseUOM <$> AT.string "cpu")
+              <|> (preview pBaseUOM <$> AT.string "vcpu-allocation")
+
+instance Show BaseUOM where
+  show = T.unpack . review pBaseUOM
+
+instance Show Prefix where
+  show = T.unpack . review pPrefixUOM
+
+instance Show UOM where
+  show = T.unpack . review pUOM
+
+instance Read UOM where
+  readsPrec _ (T.pack -> x) = maybe [] (pure . (,"")) $ x ^? pUOM
+
+instance FromJSON UOM where
+  parseJSON (String t) = maybe mzero return $ t ^? pUOM
+  parseJSON _          = mzero
+
+instance ToJSON UOM where
+  toJSON x = String $ x ^. re pUOM
+
+
+-- Convenient conversions
+
+nanosecToSec :: (UOM, Word64) -> (UOM, Word64)
+nanosecToSec (u, v)
```

I'm not sure I like the idea of this silently doing nothing when it's passed an invalid UOM - is this expected to happen in normal use? If so, can this be explained in a comment?